### ISSUE TITLE
test_grad_multiple_outs fails in DebugMode

### ISSUE THREAD

**Issue by mgermain** [Theano/Theano#3741]
The test `T_Scan.test_grad_multiple_outs` from scan_module/tests/test_scan.py is failing in _DebugMode_.

Commit where it starts to fail:
9436658304a27956ee14095cd1ad9bda39c41eb9

Commit where the error changed to the current one:
d9b0901f86c46da37e9a660f9590530d1792d702

Trace:
```
======================================================================
ERROR: test_grad_multiple_outs (theano.scan_module.tests.test_scan.T_Scan)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/Theano/theano/scan_module/tests/test_scan.py", line 1516, in test_grad_multiple_outs
    vW_in1])
  File "/home/Theano/theano/scan_module/tests/test_scan.py", line 98, in __init__
    f_x = f(*pt)
  File "/home/Theano/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
  File "/home/Theano/theano/compile/debugmode.py", line 2344, in deco
    return f()
  File "/home/Theano/theano/compile/debugmode.py", line 2269, in f
    r_vals)
  File "/home/Theano/theano/compile/debugmode.py", line 1123, in _find_bad_optimizations0
    new_graph=new_graph_str)
BadOptimization: BadOptimization Error
  Variable: id 1273097680 GpuIncSubtensor{Set;:int64:}.0
  Op GpuIncSubtensor{Set;:int64:}(GpuAllocEmpty.0, Rebroadcast{0}.0, Constant{1})
  Value Type: <type 'CudaNdarray'>
  Old Value shape, dtype, strides: (7,) float32 (1,)

  Old Value:  <CudaNdarray object at 0x464196b0>
  New Value shape, dtype, strides: (7,) float32 (1,)

  New Value:  <CudaNdarray object at 0x46420bf0>
  Reason:  local_gpu_incsubtensor
  Old Graph:
  GpuFromHost [@A] ''
   |IncSubtensor{Set;:int64:} [@B] ''
     |AllocEmpty{dtype='float32'} [@C] ''
     | |Elemwise{add,no_inplace} [@D] ''
     |   |Elemwise{switch,no_inplace} [@E] ''
     |   | |Elemwise{lt,no_inplace} [@F] ''
     |   | |Elemwise{add,no_inplace} [@G] ''
     |   | |Elemwise{sub,no_inplace} [@H] ''
     |   |TensorConstant{1} [@I]
     |Rebroadcast{0} [@J] ''
     | |DimShuffle{x} [@K] ''
     |   |HostFromGpu [@L] 'y0'
     |     |GpuFromHost [@M] ''
     |Constant{1} [@N]

  New Graph:
  GpuIncSubtensor{Set;:int64:} [@O] ''
   |GpuFromHost [@P] ''
   | |AllocEmpty{dtype='float32'} [@C] ''
   |GpuFromHost [@Q] ''
   | |Rebroadcast{0} [@J] ''
   |Constant{1} [@N]
```

**Comment by lamblin** [comment]
Could it be possible that it is not really an error, but that debugmode is comparing the content coming from the AllocEmpty, even though that part is meaningless?
We could check that by tweaking AllocEmpty so it initializes the memory in a deterministic way, and see if the error goes away. If that is the case, then we would need some mechanism to tell debugmode to ignore that part of the output of set_subtensor, or to ignore the output of set_subtensor when it ignores its input.

**Comment by nouiz** [comment]
@abergeron is this fixed?

**Comment by abergeron** [comment]
This is fixed by #4067
